tianshou/utils/space_info.py

```diff
@@ -0,0 +1,111 @@
+from collections.abc import Sequence
+from dataclasses import dataclass
+from typing import Any, Self
+
+import numpy as np
+from gymnasium import spaces
+
+from tianshou.utils.string import ToStringMixin
+
+
+@dataclass(kw_only=True)
+class ActionSpaceInfo(ToStringMixin):
+    """A data structure for storing the different attributes of the action space."""
+
+    action_shape: int | Sequence[int]
+    """The shape of the action space."""
+    min_action: float
+    """The smallest allowable action or in the continuous case the lower bound for allowable action value."""
+    max_action: float
+    """The largest allowable action or in the continuous case the upper bound for allowable action value."""
+
+    @property
+    def action_dim(self) -> int:
+        """Return the number of distinct actions an agent can take it its action space."""
+        if isinstance(self.action_shape, int):
+            return self.action_shape
+        elif isinstance(self.action_shape, Sequence) and self.action_shape:
+            return int(np.prod(self.action_shape))
+        else:
+            raise ValueError("Invalid action_shape: {self.action_shape}.")
+
+    @classmethod
+    def from_space(cls, space: spaces.Space) -> Self:
+        """Return the attributes of the action space based on the instance type of the space."""
+        if isinstance(space, spaces.Box):
+            return cls(
+                action_shape=space.shape,
+                min_action=float(np.min(space.low)),
+                max_action=float(np.max(space.high)),
+            )
+        elif isinstance(space, spaces.Discrete):
+            return cls(
+                action_shape=int(space.n),
+                min_action=float(space.start),
+                max_action=float(space.start + space.n - 1),
+            )
+        else:
+            raise ValueError(
+                f"Unsupported space type: {space.__class__}. Currently supported types are Discrete and Box.",
+            )
+
+    def _tostring_additional_entries(self) -> dict[str, Any]:
```

Comment:
I think that's unnecessary, properties are included by default

Reply:
I tried without and properties are curiously not included by default (maybe something to correct !?).

In `ToStringMixing`, `_tostring_object_info` is calling the default `_tostring_additional_entries` which returns an empty dict (cf. https://github.com/thu-ml/tianshou/blob/4756ee80ff11cd8692aef3752f35c0af60a452e8/tianshou/utils/string.py#L318 )